File: src/daytrace/sessionize.py
from __future__ import annotations

from collections.abc import Iterable
from datetime import date, timedelta
from pathlib import PurePath, PureWindowsPath

from daytrace.models import (
    ActivitySession,
    ActivitySlice,
    DayWindow,
    DiagnosticCount,
    OutcomeSignal,
    SessionBundle,
)


MERGE_GAP = timedelta(seconds=60)
BRIDGE_LIMIT = timedelta(seconds=20)
SESSION_GAP = timedelta(minutes=5)
GENERIC = {
    "chatgpt",
    "powershell",
    "new tab",
    "inbox",
    "slack",
    "outlook",
    "new repository",
    "success page",
    "installation options",
}
STANDALONE_GENERIC = {
    "chatgpt",
    "powershell",
    "new tab",
    "inbox",
    "slack",
    "outlook",
}


def _specific_key(item: ActivitySlice) -> str | None:
    for context in item.contexts:
        if context.project:
            return context.project.casefold()
        if context.url_path:
            return context.url_path.casefold()
        if context.file:
            return context.file.casefold()
    if item.title and item.title.casefold() not in GENERIC:
        return item.title.casefold()
    return None


def _path_name(value: str) -> str:
    path = PureWindowsPath(value) if "\\" in value or ":" in value else PurePath(value)
    return path.name


def _label(item: ActivitySlice) -> str:
    for context in item.contexts:
        if context.project:
            return _path_name(context.project)
        if context.url_path:
            return context.url_path.strip("/")
        if context.title:
            return context.title
    return item.title or item.app or "Unknown activity"


def _generic_name(item: ActivitySlice) -> str:
    value = (item.title or item.app or "").casefold()
    return value.removesuffix(".exe")


def _evidence_for(slices: Iterable[ActivitySlice]) -> tuple[str, ...]:
    return tuple(sorted({item_id for item in slices for item_id in item.evidence_ids}))


def _outcome_signals(slices: tuple[ActivitySlice, ...]) -> tuple[OutcomeSignal, ...]:
    signals: list[OutcomeSignal] = []
    titles = [(item.title or "").casefold() for item in slices]
    for index, title in enumerate(titles[:-1]):
        matching_later = next(
            (
                later
                for later in slices[index + 1 :]
                if any(context.url_path for context in later.contexts)
            ),
            None,
        )
        if title == "new repository" and matching_later is not None:
            path = next(
                context.url_path
                for context in matching_later.contexts
                if context.url_path
            )
            signals.append(
                OutcomeSignal(
                    "named-repository-appeared",
                    f"Named repository appeared: {path.strip('/')}",
                    _evidence_for((slices[index], matching_later)),
                )
            )
            break
    if any(title in {"success page", "checkout success"} for title in titles):
        signals.append(
            OutcomeSignal(
                "checkout-success",
                "A checkout success state appeared",
                _evidence_for(slices),
            )
        )
    if any("installer" in (item.app or "").casefold() for item in slices):
        signals.append(
            OutcomeSignal(
                "installer-started",
                "An installer was started",
                _evidence_for(slices),
            )
        )
    return tuple(signals)


def _bridge_links(items: tuple[ActivitySlice, ...]) -> set[int]:
    links: set[int] = set()
    for index in range(1, len(items) - 1):
        previous_key = _specific_key(items[index - 1])
        next_key = _specific_key(items[index + 1])
        if (
            items[index].duration_seconds < BRIDGE_LIMIT.total_seconds()
            and previous_key is not None
            and previous_key == next_key
        ):
            links.update((index, index + 1))
    return links


def _starts_new_session(
    previous: ActivitySlice,
    current: ActivitySlice,
    boundary_index: int,
    bridge_links: set[int],
) -> bool:
    if boundary_index in bridge_links:
        return False
    if current.start - previous.end > SESSION_GAP:
        return True
    previous_key = _specific_key(previous)
    current_key = _specific_key(current)
    if previous_key and current_key:
        return previous_key != current_key
    if _generic_name(previous) in STANDALONE_GENERIC:
        return True
    if _generic_name(current) in STANDALONE_GENERIC:
        return True
    if previous_key is None and current_key is None:
        return _generic_name(previous) != _generic_name(current)
    return False


def _make_session(index: int, slices: tuple[ActivitySlice, ...]) -> ActivitySession:
    active_seconds = sum(item.duration_seconds for item in slices)
    focused = [item.duration_seconds for item in slices if item.focused]
    label_source = next((item for item in slices if _specific_key(item)), slices[0])
    return ActivitySession(
        session_id=f"session-{index:03d}",
        start=min(item.start for item in slices),
        end=max(item.end for item in slices),
        active_seconds=active_seconds,
        focused_seconds=sum(focused) if focused else None,
        label=_label(label_source),
        slices=slices,
        evidence_ids=_evidence_for(slices),
        outcome_signals=_outcome_signals(slices),
    )


def sessionize(
    day: date,
    window: DayWindow,
    slices: Iterable[ActivitySlice],
    diagnostics: tuple[DiagnosticCount, ...],
) -> SessionBundle:
    ordered = tuple(sorted(slices, key=lambda item: (item.start, item.end)))
    groups: list[list[ActivitySlice]] = []
    bridge_links = _bridge_links(ordered)
    for index, item in enumerate(ordered):
        if not groups or _starts_new_session(
            ordered[index - 1], item, index, bridge_links
        ):
            groups.append([item])
        else:
            groups[-1].append(item)

    sessions = tuple(
        _make_session(index, tuple(group))
        for index, group in enumerate(groups, start=1)
    )
    focused = [item.duration_seconds for item in ordered if item.focused]
    return SessionBundle(
        day=day,
        timezone_name=window.timezone_name,
        focused_seconds=sum(focused) if focused else None,
        sessions=sessions,
        diagnostics=diagnostics,
    )

File: packages/daytrace-core/src/sessionize.ts
import type {
  ActivitySession,
  ActivitySlice,
  DayWindow,
  DiagnosticCount,
  IsoDate,
  OutcomeSignal,
  SessionBundle,
} from "./models.js";
import { durationSeconds, timestampMs } from "./time.js";

const BRIDGE_LIMIT_SECONDS = 20;
const SESSION_GAP_SECONDS = 5 * 60;
const GENERIC = new Set([
  "chatgpt", "powershell", "new tab", "inbox", "slack", "outlook",
  "new repository", "success page", "installation options",
]);
const STANDALONE_GENERIC = new Set(["chatgpt", "powershell", "new tab", "inbox", "slack", "outlook"]);

function specificKey(item: ActivitySlice): string | undefined {
  for (const signal of item.contexts) {
    if (signal.project !== undefined) return signal.project.toLocaleLowerCase();
    if (signal.urlPath !== undefined) return signal.urlPath.toLocaleLowerCase();
    if (signal.file !== undefined) return signal.file.toLocaleLowerCase();
  }
  if (item.title !== undefined && !GENERIC.has(item.title.toLocaleLowerCase())) {
    return item.title.toLocaleLowerCase();
  }
  return undefined;
}

function pathName(value: string): string {
  return value.split(/[\\/]/).filter(Boolean).at(-1) ?? value;
}

function label(item: ActivitySlice): string {
  for (const signal of item.contexts) {
    if (signal.project !== undefined) return pathName(signal.project);
    if (signal.urlPath !== undefined) return signal.urlPath.replace(/^\/+|\/+$/g, "");
    if (signal.title !== undefined) return signal.title;
  }
  return item.title ?? item.app ?? "Unknown activity";
}

function genericName(item: ActivitySlice): string {
  return (item.title ?? item.app ?? "").toLocaleLowerCase().replace(/\.exe$/, "");
}

function evidenceFor(slices: readonly ActivitySlice[]): readonly string[] {
  return [...new Set(slices.flatMap((item) => item.evidenceIds))].sort();
}

function outcomeSignals(slices: readonly ActivitySlice[]): readonly OutcomeSignal[] {
  const signals: OutcomeSignal[] = [];
  const titles = slices.map((item) => (item.title ?? "").toLocaleLowerCase());
  for (let index = 0; index < titles.length - 1; index += 1) {
    const matchingLater = slices.slice(index + 1).find(
      (later) => later.contexts.some((signal) => signal.urlPath !== undefined),
    );
    if (titles[index] === "new repository" && matchingLater !== undefined) {
      const path = matchingLater.contexts.find((signal) => signal.urlPath !== undefined)?.urlPath;
      if (path !== undefined) {
        signals.push({
          code: "named-repository-appeared",
          label: `Named repository appeared: ${path.replace(/^\/+|\/+$/g, "")}`,
          evidenceIds: evidenceFor([slices[index] as ActivitySlice, matchingLater]),
        });
      }
      break;
    }
  }
  if (titles.some((title) => title === "success page" || title === "checkout success")) {
    signals.push({ code: "checkout-success", label: "A checkout success state appeared", evidenceIds: evidenceFor(slices) });
  }
  if (slices.some((item) => (item.app ?? "").toLocaleLowerCase().includes("installer"))) {
    signals.push({ code: "installer-started", label: "An installer was started", evidenceIds: evidenceFor(slices) });
  }
  return signals;
}

function bridgeLinks(items: readonly ActivitySlice[]): Set<number> {
  const links = new Set<number>();
  for (let index = 1; index < items.length - 1; index += 1) {
    const previous = items[index - 1];
    const current = items[index];
    const next = items[index + 1];
    if (previous === undefined || current === undefined || next === undefined) continue;
    const previousKey = specificKey(previous);
    const nextKey = specificKey(next);
    if (
      durationSeconds(current.start, current.end) < BRIDGE_LIMIT_SECONDS
      && previousKey !== undefined
      && previousKey === nextKey
    ) {
      links.add(index);
      links.add(index + 1);
    }
  }
  return links;
}

function startsNewSession(
  previous: ActivitySlice,
  current: ActivitySlice,
  boundaryIndex: number,
  links: ReadonlySet<number>,
): boolean {
  if (links.has(boundaryIndex)) return false;
  if ((timestampMs(current.start) - timestampMs(previous.end)) / 1000 > SESSION_GAP_SECONDS) return true;
  const previousKey = specificKey(previous);
  const currentKey = specificKey(current);
  if (previousKey !== undefined && currentKey !== undefined) return previousKey !== currentKey;
  if (STANDALONE_GENERIC.has(genericName(previous))) return true;
  if (STANDALONE_GENERIC.has(genericName(current))) return true;
  if (previousKey === undefined && currentKey === undefined) return genericName(previous) !== genericName(current);
  return false;
}

function makeSession(index: number, slices: readonly ActivitySlice[]): ActivitySession {
  const activeSeconds = slices.reduce((total, item) => total + durationSeconds(item.start, item.end), 0);
  const focused = slices.filter((item) => item.focused).map((item) => durationSeconds(item.start, item.end));
  const labelSource = slices.find((item) => specificKey(item) !== undefined) ?? slices[0];
  if (labelSource === undefined) throw new Error("empty session");
  return {
    sessionId: `session-${String(index).padStart(3, "0")}`,
    start: slices.reduce((value, item) => timestampMs(item.start) < timestampMs(value) ? item.start : value, slices[0]!.start),
    end: slices.reduce((value, item) => timestampMs(item.end) > timestampMs(value) ? item.end : value, slices[0]!.end),
    activeSeconds,
    ...(focused.length === 0 ? {} : { focusedSeconds: focused.reduce((total, value) => total + value, 0) }),
    label: label(labelSource),
    slices,
    evidenceIds: evidenceFor(slices),
    outcomeSignals: outcomeSignals(slices),
  };
}

export function sessionize(
  day: IsoDate,
  window: DayWindow,
  slices: readonly ActivitySlice[],
  diagnostics: readonly DiagnosticCount[],
): SessionBundle {
  const ordered = [...slices].sort(
    (left, right) => timestampMs(left.start) - timestampMs(right.start) || timestampMs(left.end) - timestampMs(right.end),
  );
  const groups: ActivitySlice[][] = [];
  const links = bridgeLinks(ordered);
  ordered.forEach((item, index) => {
    const previous = ordered[index - 1];
    if (groups.length === 0 || previous === undefined || startsNewSession(previous, item, index, links)) {
      groups.push([item]);
    } else {
      groups.at(-1)?.push(item);
    }
  });
  const sessions = groups.map((group, index) => makeSession(index + 1, group));
  const focused = ordered.filter((item) => item.focused).map((item) => durationSeconds(item.start, item.end));
  return {
    day,
    timezoneName: window.timezoneName,
    ...(focused.length === 0 ? {} : { focusedSeconds: focused.reduce((total, value) => total + value, 0) }),
    sessions,
    diagnostics,
  };
}
